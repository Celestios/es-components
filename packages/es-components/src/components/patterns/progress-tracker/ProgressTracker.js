import React from 'react';
import { withTheme } from 'styled-components';
import PropTypes from 'prop-types';
import {
  ProgressContainer,
  ProgressItem
} from './progress-tracker-subcomponents';

const getMappedSteps = (steps, onPastStepClicked, onFutureStepClicked) => {
  let isPastStep = true;
  return steps.map((step, index) => {
    isPastStep = !(step.active || !isPastStep);
    return (
      <ProgressItem
        /* eslint-disable react/no-array-index-key */
        key={`${index}-${step.label}`}
        /* eslint-enable */
        active={step.active}
        isPastStep={isPastStep}
        numberOfSteps={steps.length}
        onPastStepClicked={() => onPastStepClicked(index)}
        canClickFutureStep={
          onFutureStepClicked !== null && onFutureStepClicked !== undefined
        }
        onFutureStepClicked={() => onFutureStepClicked(index)}
        label={step.label}
      />
    );
  });
};

const getIndexOfActiveStep = steps => steps.findIndex(step => step.active);

function ProgressTracker({ steps, onPastStepClicked, onFutureStepClicked }) {
  return (
    <ProgressContainer
      data-testid="progress-tracker"
      activeStepIndex={getIndexOfActiveStep(steps)}
      numberOfSteps={steps.length}
    >
      {getMappedSteps(steps, onPastStepClicked, onFutureStepClicked)}
    </ProgressContainer>
  );
}

ProgressTracker.propTypes = {
  steps: PropTypes.arrayOf(
    PropTypes.shape({
      active: PropTypes.bool,
      label: PropTypes.string
    })
  ),
  onPastStepClicked: PropTypes.func,
  onFutureStepClicked: PropTypes.func
};

ProgressTracker.defaultProps = {
  steps: [],
  onPastStepClicked() {},
  onFutureStepClicked: null
};

export default withTheme(ProgressTracker);
