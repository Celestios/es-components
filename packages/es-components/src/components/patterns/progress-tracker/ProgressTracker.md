Initial Progress Tracker state.
```
<ProgressTracker steps={[
  { active: true, label: "Item One" },
  { active: false, label: "Item Two" },
  { active: false, label: "Item Three" },
  { active: false, label: "Item Four" },
]} />
```

Step Two Active:
```
<ProgressTracker steps={[
  { active: false, label: "Item One" },
  { active: true, label: "Item Two" },
  { active: false, label: "Item Three" },
  { active: false, label: "Item Four" },
]} />
```

Step Three Active:
```
<ProgressTracker steps={[
  { active: false, label: "Item One" },
  { active: false, label: "Item Two" },
  { active: true, label: "Item Three" },
  { active: false, label: "Item Four" },
]} />
```

Step Four Active:
```
<ProgressTracker steps={[
  { active: false, label: "Item One" },
  { active: false, label: "Item Two" },
  { active: false, label: "Item Three" },
  { active: true, label: "Item Four" },
]} />
```

Many steps:
```
<ProgressTracker steps={[
  { active: false, label: "Item One" },
  { active: false, label: "Item Two" },
  { active: false, label: "Item Three" },
  { active: false, label: "Item Four" },
  { active: false, label: "Item Five" },
  { active: false, label: "Item Six" },
  { active: true, label: "Item Seven" },
  { active: false, label: "Item Eight" },
]} />
```

Responding when a past step is clicked. The intended behavior is that, if
you go backward in history, you _must_ complete the new current step and any
steps after that. That being the case, if you go backwards, the only way to
continue forward is to complete the steps back to where you previously were. In
this example, we've provided a "Reset" button to allow you to set the component
to its original state.
```javascript

const initialSteps = [
  { active: false, label: "Item One" },
  { active: false, label: "Item Two" },
  { active: false, label: "Item Three" },
  { active: false, label: "Item Four" },
  { active: false, label: "Item Five" },
  { active: false, label: "Item Six" },
  { active: true, label: "Item Seven" },
  { active: false, label: "Item Eight" },
];
const [steps, setSteps] = React.useState(initialSteps);

<div>
  <ProgressTracker
    steps={steps}
    onPastStepClicked={(newIndex) => setSteps(
       steps.map((step, index) => ({ ...step, active: index === newIndex }))
    )}
  />
  <button onClick={() => setSteps(initialSteps)}>Reset</button>
</div>
```
