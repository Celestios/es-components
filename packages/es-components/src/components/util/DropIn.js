import React from 'react';
import PropTypes from 'prop-types';
import Transition from 'react-transition-group/Transition';

const DropIn = ({ children, duration, ...otherProps }) => {
  const transitionStyles = {
    opacity: 0,
    transform: `translate(0, -25%)`,
    transition: `transform ${duration}ms ease-out, opacity ${duration}ms linear`
  };

  const stateStyles = {
    entering: { opacity: 1, transform: `translate(0, 0)` },
    entered: { opacity: 1, transform: `translate(0, 0)` }
  };

  return (
    <Transition {...otherProps} timeout={duration}>
      {(state, innerProps) =>
        React.cloneElement(children, {
          ...innerProps,
          style: {
            ...children.props.style,
            ...transitionStyles,
            ...stateStyles[state]
          }
        })
      }
    </Transition>
  );
};

DropIn.propTypes = {
  children: PropTypes.any.isRequired,
  duration: PropTypes.number
};

DropIn.defaultProps = {
  duration: 300
};

export default DropIn;
