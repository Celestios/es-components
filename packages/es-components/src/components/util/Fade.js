import React from 'react';
import PropTypes from 'prop-types';
import Transition from 'react-transition-group/Transition';

const Fade = ({ children, duration, opacity, ...otherProps }) => {
  const transitionStyles = {
    transition: `opacity ${duration}ms linear`,
    opacity: 0
  };

  const stateStyles = {
    entering: { opacity },
    entered: { opacity }
  };

  return (
    <Transition {...otherProps} timeout={duration}>
      {(state, innerProps) =>
        React.cloneElement(children, {
          ...innerProps,
          style: {
            ...children.props.style,
            ...transitionStyles,
            ...stateStyles[state]
          }
        })
      }
    </Transition>
  );
};

Fade.propTypes = {
  children: PropTypes.any.isRequired,
  duration: PropTypes.number,
  opacity: PropTypes.number
};

Fade.defaultProps = {
  duration: 150,
  opacity: 1
};

export default Fade;
