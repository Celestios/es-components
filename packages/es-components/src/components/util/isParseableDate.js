import parse from 'date-fns/parse';
import isValid from 'date-fns/isValid';

export default function isParseableDate(date) {
  const parsedDate = parse(date);
  return isValid(parsedDate);
}
