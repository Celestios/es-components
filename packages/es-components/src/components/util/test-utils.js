// eslint-env jest
import React from 'react';
import { ThemeProvider } from 'styled-components';
import viaTheme from 'es-components-via-theme';
import { render } from '@testing-library/react';

export function renderWithTheme(component) {
  return render(
    <ThemeProvider theme={viaTheme}>
      {React.cloneElement(component)}
    </ThemeProvider>
  );
}
