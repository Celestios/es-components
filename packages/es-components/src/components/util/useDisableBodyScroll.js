import { useEffect } from 'react';

export const useDisableBodyScroll = shouldDisableScroll => {
  useEffect(() => {
    const newStyle = 'overflow: hidden;';
    const getCurrentStyle = () => document.body.getAttribute('style') || '';
    const disableScroll = () =>
      document.body.setAttribute('style', `${getCurrentStyle()} ${newStyle}`);
    const enableScroll = () =>
      document.body.setAttribute(
        'style',
        getCurrentStyle().replace(newStyle, '')
      );

    if (shouldDisableScroll) {
      disableScroll();
    } else {
      enableScroll();
    }

    return enableScroll;
  }, [shouldDisableScroll]);
};
