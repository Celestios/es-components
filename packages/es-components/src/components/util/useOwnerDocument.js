import { useState, useCallback } from 'react';

export default function useOwnerDocument(currentDocument) {
  const [ownerDocument, setOwnerDocument] = useState(currentDocument);
  const ownerDocumentRef = useCallback(node => {
    if (!node) return;
    const targetDocument =
      node.ownerDocument?.body || node.ownerDocument || currentDocument;
    if (ownerDocument !== targetDocument) setOwnerDocument(targetDocument);
  });
  return [ownerDocument, ownerDocumentRef];
}
