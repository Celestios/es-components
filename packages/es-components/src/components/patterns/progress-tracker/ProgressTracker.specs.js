/* eslint-env jest */

import React from 'react';
import { cleanup } from '@testing-library/react';

import { renderWithTheme } from '../../util/test-utils';
import ProgressTracker from './ProgressTracker';

function getInstance(props) {
  return renderWithTheme(<ProgressTracker {...props} />);
}

it('builds a ProgressTracker', () => {
  const stepsToTest = [
    [
      { label: 'step 1', active: true },
      { label: 'step 2', active: false },
      { label: 'step 3', active: false }
    ],
    [
      { label: 'step 1', active: false },
      { label: 'step 2', active: true },
      { label: 'step 3', active: false }
    ],
    [
      { label: 'step 1', active: false },
      { label: 'step 2', active: false },
      { label: 'step 3', active: true }
    ]
  ];
  stepsToTest.forEach(steps => {
    const { getByTestId } = getInstance({ steps });
    expect(getByTestId('progress-tracker')).toMatchSnapshot();
    cleanup();
  });
});
