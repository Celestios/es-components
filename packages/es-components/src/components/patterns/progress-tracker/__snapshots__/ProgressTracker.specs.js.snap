// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`builds a ProgressTracker 1`] = `
<ol
  class="sc-bdVaJa cMKOYT"
  data-testid="progress-tracker"
>
  <li
    class="sc-bZQynM fRYQzq"
  >
    <button
      class="sc-bwzfXH sc-EHOje enaMrX"
      disabled=""
    >
      <span>
        step 1
      </span>
    </button>
  </li>
  <li
    class="sc-bZQynM fRYQzq"
  >
    <button
      class="sc-bwzfXH fZOxh"
      disabled=""
    >
      <span>
        step 2
      </span>
    </button>
  </li>
  <li
    class="sc-bZQynM fRYQzq"
  >
    <button
      class="sc-bwzfXH fZOxh"
      disabled=""
    >
      <span>
        step 3
      </span>
    </button>
  </li>
</ol>
`;

exports[`builds a ProgressTracker 2`] = `
<ol
  class="sc-bdVaJa cMKOYT"
  data-testid="progress-tracker"
>
  <li
    class="sc-bZQynM fRYQzq"
  >
    <button
      class="sc-bwzfXH sc-EHOje enaMrX"
      disabled=""
    >
      <span>
        step 1
      </span>
    </button>
  </li>
  <li
    class="sc-bZQynM fRYQzq"
  >
    <button
      class="sc-bwzfXH fZOxh"
      disabled=""
    >
      <span>
        step 2
      </span>
    </button>
  </li>
  <li
    class="sc-bZQynM fRYQzq"
  >
    <button
      class="sc-bwzfXH fZOxh"
      disabled=""
    >
      <span>
        step 3
      </span>
    </button>
  </li>
</ol>
`;

exports[`builds a ProgressTracker 3`] = `
<ol
  class="sc-bdVaJa cMKOYT"
  data-testid="progress-tracker"
>
  <li
    class="sc-bZQynM fRYQzq"
  >
    <button
      class="sc-bwzfXH sc-EHOje enaMrX"
      disabled=""
    >
      <span>
        step 1
      </span>
    </button>
  </li>
  <li
    class="sc-bZQynM fRYQzq"
  >
    <button
      class="sc-bwzfXH fZOxh"
      disabled=""
    >
      <span>
        step 2
      </span>
    </button>
  </li>
  <li
    class="sc-bZQynM fRYQzq"
  >
    <button
      class="sc-bwzfXH fZOxh"
      disabled=""
    >
      <span>
        step 3
      </span>
    </button>
  </li>
</ol>
`;
